File: src/main/java/io/github/slimefunguguproject/bump/implementation/tasks/WeaponProjectileTask.java
package io.github.slimefunguguproject.bump.implementation.tasks;

import java.util.HashMap;
import java.util.Iterator;
import java.util.Map;

import javax.annotation.Nonnull;

import com.google.common.base.Preconditions;

import org.bukkit.entity.Projectile;

import io.github.slimefunguguproject.bump.implementation.Bump;
import io.github.thebusybiscuit.slimefun4.implementation.Slimefun;

import net.guizhanss.guizhanlib.common.Scheduler;

/**
 * The {@link WeaponProjectileTask} is responsible for tracking
 * {@link Projectile} fired from Bump weapons.
 *
 * @author ybw0014
 */
@SuppressWarnings("ConstantConditions")
public final class WeaponProjectileTask implements Runnable {

    private static WeaponProjectileTask instance;

    // This map records each projectile's spawn time
    private final Map<Projectile, Integer> projectileMap = new HashMap<>();
    private final int duration;

    public WeaponProjectileTask(int duration) {
        instance = this;
        this.duration = duration;
    }

    /**
     * This method starts this task
     */
    public static void start() {
        int duration = Bump.getRegistry().getConfig().getInt("weapons.projectile-duration", 0, 60);
        if (duration > 0) {
            Scheduler.repeat(Slimefun.getTickerTask().getTickRate(), new WeaponProjectileTask(duration));
        }
    }

    /**
     * This method will add {@link Projectile} to tracking list.
     *
     * @param projectile the {@link Projectile} to be added.
     */
    public static void track(@Nonnull Projectile projectile) {
        Preconditions.checkArgument(projectile != null, "Projectile cannot not be null. How?");
        instance.trackProjectile(projectile);
    }

    @Override
    public void run() {
        int currentTick = Bump.getSlimefunTickCount();

        Iterator<Map.Entry<Projectile, Integer>> it = projectileMap.entrySet().iterator();
        while (it.hasNext()) {
            Map.Entry<Projectile, Integer> entry = it.next();
            if (entry.getValue() + duration < currentTick) {
                Projectile projectile = entry.getKey();
                if (projectile.isValid()) {
                    projectile.remove();
                }
                it.remove();
            }
        }
    }

    private void trackProjectile(Projectile projectile) {
        projectileMap.put(projectile, Bump.getSlimefunTickCount());
    }
}

File: src/main/kotlin/io/github/slimefunguguproject/bump/implementation/tasks/WeaponProjectileTask.kt
package io.github.slimefunguguproject.bump.implementation.tasks

import io.github.slimefunguguproject.bump.Bump
import io.github.slimefunguguproject.bump.utils.constant.Keys
import io.github.thebusybiscuit.slimefun4.implementation.Slimefun
import io.github.thebusybiscuit.slimefun4.libraries.dough.data.persistent.PersistentDataAPI
import org.bukkit.entity.Projectile

object WeaponProjectileTask : Runnable {
    private val projectileMap: MutableMap<Projectile, Int> = HashMap()

    /**
     * This method starts this task
     */
    fun start() {
        val duration = Bump.configService.weaponProjectileDuration
        if (duration > 0) {
            Bump.scheduler().repeat(Slimefun.getTickerTask().tickRate, this)
        }
    }

    /**
     * This method will add [Projectile] to tracking list.
     *
     * @param projectile the [Projectile] to be added.
     */
    fun track(projectile: Projectile) {
        projectileMap[projectile] = Bump.sfTickCount()
        PersistentDataAPI.setBoolean(projectile, Keys.PROJECTILE, true)
    }

    override fun run() {
        val currentTick = Bump.sfTickCount()

        val it = projectileMap.entries.iterator()
        while (it.hasNext()) {
            val entry = it.next()
            if (entry.value + Bump.configService.weaponProjectileDuration < currentTick) {
                val projectile = entry.key
                if (projectile.isValid) {
                    projectile.remove()
                }
                it.remove()
            }
        }
    }


}

